File: bindings/elixir/lib/rockbox/dsp.ex
defmodule Rockbox.Dsp do
  @moduledoc """
  The DSP pipeline: EQ, tone, surround, compressor, ReplayGain, resampler.

  A DSP handle wraps the process-wide audio DSP singleton, so only one should
  be alive at a time. The handle is a NIF resource — it is freed by the BEAM
  garbage collector, no explicit close needed.

  ReplayGain `mode` here uses the *DSP-native* values (see `t:rg_mode/0`):
  `0` track, `1` album, `2` shuffle, `3` off.
  """

  @typedoc "Opaque DSP handle (a NIF resource)."
  @opaque t :: reference()

  @typedoc "DSP ReplayGain mode: 0 track, 1 album, 2 shuffle, 3 off."
  @type rg_mode :: 0..3

  @doc "Create a DSP for interleaved S16LE stereo at `sample_rate` Hz."
  @spec new(pos_integer()) :: t()
  def new(sample_rate), do: :rockbox_ffi_nif.dsp_new(sample_rate)

  @spec set_input_frequency(t(), pos_integer()) :: :ok
  def set_input_frequency(d, hz), do: nilok(:rockbox_ffi_nif.dsp_set_input_frequency(d, hz))

  @spec flush(t()) :: :ok
  def flush(d), do: nilok(:rockbox_ffi_nif.dsp_flush(d))

  @spec eq_enable(t(), boolean()) :: :ok
  def eq_enable(d, enable), do: nilok(:rockbox_ffi_nif.dsp_eq_enable(d, enable))

  @doc "Configure one EQ band (0..9). Band 0 low shelf, 9 high shelf."
  @spec set_eq_band(t(), 0..9, integer(), number(), number()) :: :ok
  def set_eq_band(d, band, cutoff_hz, q, gain_db),
    do: nilok(:rockbox_ffi_nif.dsp_set_eq_band(d, band, cutoff_hz, q / 1, gain_db / 1))

  @spec set_eq_precut(t(), number()) :: :ok
  def set_eq_precut(d, db), do: nilok(:rockbox_ffi_nif.dsp_set_eq_precut(d, db / 1))

  @spec set_tone(t(), integer(), integer()) :: :ok
  def set_tone(d, bass_db, treble_db), do: nilok(:rockbox_ffi_nif.dsp_set_tone(d, bass_db, treble_db))

  @spec set_tone_cutoffs(t(), integer(), integer()) :: :ok
  def set_tone_cutoffs(d, bass_hz, treble_hz),
    do: nilok(:rockbox_ffi_nif.dsp_set_tone_cutoffs(d, bass_hz, treble_hz))

  @spec set_surround(t(), integer(), integer(), integer(), integer()) :: :ok
  def set_surround(d, delay_ms, balance, fx1, fx2),
    do: nilok(:rockbox_ffi_nif.dsp_set_surround(d, delay_ms, balance, fx1, fx2))

  @spec set_channel_config(t(), integer()) :: :ok
  def set_channel_config(d, mode), do: nilok(:rockbox_ffi_nif.dsp_set_channel_config(d, mode))

  @spec set_stereo_width(t(), integer()) :: :ok
  def set_stereo_width(d, pct), do: nilok(:rockbox_ffi_nif.dsp_set_stereo_width(d, pct))

  @spec set_compressor(t(), integer(), integer(), integer(), integer(), integer(), integer()) :: :ok
  def set_compressor(d, threshold, makeup_gain, ratio, knee, release_time, attack_time),
    do:
      nilok(
        :rockbox_ffi_nif.dsp_set_compressor(
          d, threshold, makeup_gain, ratio, knee, release_time, attack_time
        )
      )

  @doc "ReplayGain mode (see `t:rg_mode/0`), clipping prevention, preamp in dB."
  @spec set_replaygain(t(), rg_mode(), boolean(), number()) :: :ok
  def set_replaygain(d, mode, noclip, preamp_db),
    do: nilok(:rockbox_ffi_nif.dsp_set_replaygain(d, mode, noclip, preamp_db / 1))

  @doc """
  Per-track gains in plain dB / peaks as linear amplitude. Pass `nil` for an
  absent tag.
  """
  @spec set_replaygain_gains(
          t(),
          number() | nil,
          number() | nil,
          number() | nil,
          number() | nil
        ) :: :ok
  def set_replaygain_gains(d, track_gain_db, album_gain_db, track_peak, album_peak),
    do:
      nilok(
        :rockbox_ffi_nif.dsp_set_replaygain_gains(
          d, f(track_gain_db), f(album_gain_db), f(track_peak), f(album_peak)
        )
      )

  @doc "Native Q7.24 factors (the `raw_*` metadata fields); 0 = not tagged."
  @spec set_replaygain_gains_raw(t(), integer(), integer(), integer(), integer()) :: :ok
  def set_replaygain_gains_raw(d, track_gain, album_gain, track_peak, album_peak),
    do:
      nilok(
        :rockbox_ffi_nif.dsp_set_replaygain_gains_raw(
          d, track_gain, album_gain, track_peak, album_peak
        )
      )

  @doc """
  Run interleaved stereo S16 samples through the pipeline. Accepts and returns
  a binary of little-endian int16 samples (use `samples_to_binary/1` /
  `binary_to_samples/1` to convert lists).
  """
  @spec process(t(), binary()) :: binary()
  def process(d, pcm) when is_binary(pcm), do: :rockbox_ffi_nif.dsp_process(d, pcm)

  @doc "Convert a list of int16 samples to a little-endian binary."
  @spec samples_to_binary([integer()]) :: binary()
  def samples_to_binary(samples),
    do: for(s <- samples, into: <<>>, do: <<s::16-little-signed>>)

  @doc "Convert a little-endian int16 binary to a list of samples."
  @spec binary_to_samples(binary()) :: [integer()]
  def binary_to_samples(bin), do: for(<<s::16-little-signed <- bin>>, do: s)

  # nil (absent) or a float; keep nil so the NIF maps it to NaN.
  defp f(nil), do: nil
  defp f(n), do: n / 1

  defp nilok(nil), do: :ok
  defp nilok(other), do: other
end

File: bindings/go/dsp.go
package rockbox

import (
	"errors"
	"math"
	"unsafe"
)

// Dsp is the Rockbox DSP pipeline: EQ, tone, surround, compressor, ReplayGain,
// and resampler.
//
// The underlying dsp_config is a process-wide singleton, so only one Dsp may
// exist at a time and it must be used from one goroutine. Call [Dsp.Close] when
// done.
type Dsp struct {
	ptr uintptr
}

// NewDsp creates a Dsp configured for the given output sample rate (Hz).
func NewDsp(sampleRate uint32) (*Dsp, error) {
	ptr := rbDspNew(sampleRate)
	if ptr == 0 {
		return nil, errors.New("rockbox: rb_dsp_new returned NULL")
	}
	return &Dsp{ptr: ptr}, nil
}

// Close frees the native DSP. It is safe to call more than once.
func (d *Dsp) Close() {
	if d.ptr == 0 {
		return
	}
	rbDspFree(d.ptr)
	d.ptr = 0
}

// SetInputFrequency sets the sample rate (Hz) of the audio fed to Process, so
// the resampler can convert it to the output rate.
func (d *Dsp) SetInputFrequency(hz uint32) { rbDspSetInputFrequency(d.ptr, hz) }

// Flush clears the internal resampler / filter state.
func (d *Dsp) Flush() { rbDspFlush(d.ptr) }

// EqEnable turns the parametric EQ on or off.
func (d *Dsp) EqEnable(enable bool) { rbDspEqEnable(d.ptr, enable) }

// SetEqBand configures one EQ band (0..=9; band 0 is the low shelf, 9 the high
// shelf). q is the Q factor and gainDb the gain in dB, both in plain units —
// the binding applies Rockbox's tenths scaling internally.
func (d *Dsp) SetEqBand(band int, cutoffHz int32, q, gainDb float32) {
	rbDspSetEqBand(d.ptr, uint64(band), cutoffHz, q, gainDb)
}

// SetEqPrecut sets the EQ pre-cut in dB.
func (d *Dsp) SetEqPrecut(db float32) { rbDspSetEqPrecut(d.ptr, db) }

// SetTone sets the bass and treble shelves in dB.
func (d *Dsp) SetTone(bassDb, trebleDb int32) { rbDspSetTone(d.ptr, bassDb, trebleDb) }

// SetToneCutoffs sets the bass and treble shelf cutoff frequencies (Hz).
func (d *Dsp) SetToneCutoffs(bassHz, trebleHz int32) { rbDspSetToneCutoffs(d.ptr, bassHz, trebleHz) }

// SetSurround configures the surround effect.
func (d *Dsp) SetSurround(delayMs, balance, fx1, fx2 int32) {
	rbDspSetSurround(d.ptr, delayMs, balance, fx1, fx2)
}

// SetChannelConfig selects the channel routing (see [ChannelConfig]).
func (d *Dsp) SetChannelConfig(mode ChannelConfig) { rbDspSetChannelConfig(d.ptr, int32(mode)) }

// SetStereoWidth sets the stereo width as a percentage.
func (d *Dsp) SetStereoWidth(percent int32) { rbDspSetStereoWidth(d.ptr, percent) }

// SetCompressor configures the dynamic-range compressor.
func (d *Dsp) SetCompressor(threshold, makeupGain, ratio, knee, releaseTime, attackTime int32) {
	rbDspSetCompressor(d.ptr, threshold, makeupGain, ratio, knee, releaseTime, attackTime)
}

// SetReplaygain selects the ReplayGain mode (see [DspReplayGainMode]). When
// noclip is true, gains that would clip are reduced; preampDb is applied on top.
func (d *Dsp) SetReplaygain(mode DspReplayGainMode, noclip bool, preampDb float32) {
	rbDspSetReplaygain(d.ptr, int32(mode), noclip, preampDb)
}

// SetReplaygainGains sets the per-track ReplayGain values: gains in plain dB
// and peaks as linear amplitude (1.0 = full scale). Pass a nil pointer for any
// absent tag (mapped to the ABI's NaN sentinel). Use [Opt] to build pointers.
func (d *Dsp) SetReplaygainGains(trackGainDb, albumGainDb, trackPeak, albumPeak *float32) {
	rbDspSetReplaygainGains(d.ptr, optF32(trackGainDb), optF32(albumGainDb), optF32(trackPeak), optF32(albumPeak))
}

// SetReplaygainGainsRaw sets the native Q7.24 linear factors (the Raw* fields
// of [ReplayGain] from [Metadata.Read]); 0 means "not tagged".
func (d *Dsp) SetReplaygainGainsRaw(trackGain, albumGain, trackPeak, albumPeak int64) {
	rbDspSetReplaygainGainsRaw(d.ptr, trackGain, albumGain, trackPeak, albumPeak)
}

// Process runs interleaved stereo S16 samples through the pipeline and returns
// the processed samples. The output length may differ from the input (the
// resampler buffers). input must have even length (L/R pairs).
func (d *Dsp) Process(input []int16) ([]int16, error) {
	if len(input)%2 != 0 {
		return nil, errors.New("rockbox: input must be interleaved stereo (even length)")
	}
	if len(input) == 0 {
		return nil, nil
	}
	var outLen uint64
	outPtr := rbDspProcess(d.ptr, unsafe.Pointer(&input[0]), uint64(len(input)), unsafe.Pointer(&outLen))
	if outPtr == 0 || outLen == 0 {
		return nil, nil
	}
	src := unsafe.Slice((*int16)(unsafe.Pointer(outPtr)), outLen)
	out := make([]int16, outLen)
	copy(out, src)
	rbBufferFree(outPtr, outLen)
	return out, nil
}

// Opt returns a pointer to v, for the optional *float32 arguments of
// [Dsp.SetReplaygainGains].
func Opt(v float32) *float32 { return &v }

// optF32 maps a nil pointer to NaN (the ABI's "tag absent" sentinel).
func optF32(v *float32) float32 {
	if v == nil {
		return float32(math.NaN())
	}
	return *v
}
